### PR TITLE
Threadprivate fix

### LINKED ISSUE #254: Race reported on OpenMP threadprivate variable
**Describe the bug**
in dataracebench case 171 the variable `x` is marked as threadprivate, but we still report a race on `x`.

**Expected behavior**
Races should not be reported on threadprivate variables.

**To Reproduce**
Run the tool on `tests/data/integration/dataracebench/DRB171-threadprivate3-orig-no.ll`.

Observe the following race:

```shell
DRB171-threadprivate3-orig-no.c:26:10 DRB171-threadprivate3-orig-no.c:28:11
          store double -1.000000e+00, double* %17, align 8, !dbg !65, !tbaa !66, !noalias !34
          %19 = load i64, i64* %18, align 16, !dbg !71, !tbaa !66, !noalias !34
```

**Additional context**
in the IR `%17` is a GEP on `@x` and `%18` is bitcasted `@x`:

```llvm
%17 = getelementptr [20 x double], [20 x double]* @x, i32 0, i64 %indvars.iv.i,
%18 = bitcast [20 x double]* @x to i64*,
```
I also confirmed that `@x` is threadlocal:

```llvm
@x = internal thread_local global ...
```

### PR DESCRIPTION
**Pull request by BradSwain**
Closes #254 

Add some logic to the threadprivate check to see if the pointer being operated on flows from a global threadlocal object.

### CODE REVIEW

**Review by funemy — APPROVED** [review]
Looks good to me for this _specific_ case. So I approved.

Would it make more sense to iterate over use-def chain?

**Comment by BradSwain** [comment]
I think no, because we only want to mark stuff as thread local if the pointer being operated on must point to a global.

If there is something like:

```cpp
int threadlocal; // this is thread local

// ...
#pragma omp parallel
{
  int x = threadlocal + 1;
}
```

Following the entire use-def chain will result in `x` being marked as threadlocal because `x` "uses" `threadlocal`.

I think we just want  to follow llvm instructions that copy the pointer. 
I think load and GEP are the only two that do this, but I'm not entirely sure.

**Review by funemy — CHANGES_REQUESTED** [review]
According to your issue, you should also handle `StoreInst`?

**Comment by BradSwain** [comment]
Updating to move this check to happen during analysis phase and use PTA.

File: src/Analysis/ThreadLocalAnalysis.cpp [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+using namespace race;
+
+bool ThreadLocalAnalysis::isThreadLocalAccess(const MemAccessEvent *write, const MemAccessEvent *other) {
+  // Get the intersection of the pts to set and
+  // check that each obj in the intersection is a thread local value
+
+  // This handles cases where given to accesses with pts to sets like
+  // write: { O1, O2 }
+  // other: { O1, O3 }
+  // where O1 is the onlt thread local object
```
Comment (funemy) [inline src/Analysis/ThreadLocalAnalysis.cpp:23]: "onlt" -> "only"

File: src/Analysis/ThreadLocalAnalysis.cpp [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+  // We should not report a race because the only possible
+  // shared object is thread local.
+
+  auto writePtsTo = write->getAccessedMemory();
+  auto otherPtsTo = other->getAccessedMemory();
+
+  std::sort(writePtsTo.begin(), writePtsTo.end());
+  std::sort(otherPtsTo.begin(), otherPtsTo.end());
```
Comment (funemy) [inline src/Analysis/ThreadLocalAnalysis.cpp:32]: Why sorting is needed first?

File: src/Analysis/ThreadLocalAnalysis.cpp [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+  std::vector<const pta::ObjTy *> shared;
+  std::set_intersection(writePtsTo.begin(), writePtsTo.end(), otherPtsTo.begin(), otherPtsTo.end(),
+                        std::back_inserter(shared));
+
+  return std::all_of(shared.begin(), shared.end(), [](const pta::ObjTy *obj) {
+    auto const val = obj->getValue();
+
+    auto const global = llvm::dyn_cast_or_null<llvm::GlobalVariable>(val);
+    llvm::outs() << *val << " " << (global && global->isThreadLocal()) << "\n";
```
Comment (funemy) [inline src/Analysis/ThreadLocalAnalysis.cpp:42]: do we want to keep this debugging log?

File: src/IR/Builder.cpp [hunk 4]
```diff
@@ -26,7 +26,10 @@ extern llvm::cl::opt<bool> DEBUG_PTA;
 
 namespace {
 
+// return true if the operand of inst must be a thread local object
 bool hasThreadLocalOperand(const llvm::Instruction *inst) {
```
Comment (funemy) [inline src/IR/Builder.cpp:30]: is this function used anywhere?

**Review by funemy — CHANGES_REQUESTED** [review]

File: src/Analysis/ThreadLocalAnalysis.cpp (hunk 3 above)
Comment (BradSwain) [reply src/Analysis/ThreadLocalAnalysis.cpp:42]: oops, lol. we do not.

File: src/Analysis/ThreadLocalAnalysis.cpp (hunk 2 above)
Comment (BradSwain) [reply src/Analysis/ThreadLocalAnalysis.cpp:32]: I think set_intersection requires the list to be sorted.

File: src/IR/Builder.cpp (hunk 4 above)
Comment (BradSwain) [reply src/IR/Builder.cpp:30]: yeah. Used in the load/store branches of function summary builder

File: src/Analysis/ThreadLocalAnalysis.cpp (hunk 2 above)
Comment (funemy) [reply src/Analysis/ThreadLocalAnalysis.cpp:32]: I see, I forgot that lol

Comment (BradSwain) [reply src/Analysis/ThreadLocalAnalysis.cpp:32]: I double checked
> Elements are compared using operator< and the ranges must be sorted with respect to the same.
https://en.cppreference.com/w/cpp/algorithm/set_intersection

**Review by funemy — APPROVED** [review]
LGTM

**Pull request merged**